File: 17/pset17b.py
import copy

SPACELIMIT = 31
MIDDLE = SPACELIMIT // 2

def main():
    with open("input.txt", "r") as f:
        rawinput = f.read().splitlines()

    space = []
    for a in range(SPACELIMIT):
        superslice = []
        for i in range(SPACELIMIT):
            slice = []
            for j in range(SPACELIMIT):
                row = []
                for k in range(SPACELIMIT):
                    row.append(".")
                slice.append(row)
            superslice.append(slice)
        space.append(superslice)

    for index, i in enumerate(rawinput):
        for indexj, j in enumerate(i):
            space[MIDDLE][MIDDLE][index - len(rawinput) // 2 + MIDDLE][indexj- len(rawinput) // 2 + MIDDLE] = j
            
    for iteration in range(6):
        space2 = copy.deepcopy(space)
        for indexa, a in enumerate(space): # For every 3d superslice
            for indexi, i in enumerate(a): # For every 2d slice
                for indexj, j in enumerate(i): # For every row
                    for indexk, k in enumerate(j): # For every cell
                        if k == "#" and countaround(space2, indexa, indexi, indexj, indexk) not in [2, 3]:
                            space[indexa][indexi][indexj][indexk] = "."
                        if k == "." and countaround(space2, indexa, indexi, indexj, indexk) == 3:
                            space[indexa][indexi][indexj][indexk] = "#"

    count = 0
    for a in space:
        for i in a:
            for j in i:
                for k in j:
                    if k == "#":
                        count += 1
                
    print(count)


def countaround (space, superslice, slice, row, col):
    count = 0
    if superslice in [0, SPACELIMIT - 1] or slice in [0, SPACELIMIT - 1] or row in [0, SPACELIMIT - 1] or col in [0, SPACELIMIT - 1]:
        return 0;
    for a in range(superslice - 1, superslice + 2):
        for i in range(slice - 1, slice + 2):
            for j in range(row - 1, row + 2):
                for k in range(col - 1, col + 2):
                    if space[a][i][j][k] == "#":
                        count += 1
    
    if space[superslice][slice][row][col] == "#":
        count -= 1
    return count
        


if __name__ == "__main__":
    main()

File: 17/pset17b.cpp
#include <iostream>
#include <fstream>
#include <vector>
#include <cmath>

using namespace std;

const int SPACELIMIT = 31;
const int MIDDLE = floor(SPACELIMIT / 2);

int count_around(char[SPACELIMIT][SPACELIMIT][SPACELIMIT][SPACELIMIT], int, int, int, int);

int main() {
    
    char space[SPACELIMIT][SPACELIMIT][SPACELIMIT][SPACELIMIT];
    
    // Initialise space
    for (int a = 0; a < SPACELIMIT; ++a)
        for (int i = 0; i < SPACELIMIT; ++i)
            for (int j = 0; j < SPACELIMIT; ++j)
                for (int k = 0; k < SPACELIMIT; ++k)
                    space[a][i][j][k] = '.';

    // Place input into space
    std::vector<std::string> raw_input;
    ifstream myfile("input.txt");
    std::string line;
    if (myfile.is_open()) {
        while (getline(myfile, line)) {
            raw_input.push_back(line);
        }
        myfile.close();
    }

    int start_index = MIDDLE - floor(raw_input.size() / 2);
    
    for (int i = 0; i < raw_input.size(); ++i) 
        for (int j = 0; j < raw_input[i].length(); ++j) 
            space[MIDDLE][MIDDLE][i + start_index][j + start_index] = raw_input[i].at(j);

    // Start the game of life
    for (int iteration = 0; iteration < 6; ++iteration) {

        // Copy the space
        char space2[SPACELIMIT][SPACELIMIT][SPACELIMIT][SPACELIMIT];
        for (int a = 0; a < SPACELIMIT; ++a)
            for (int i = 0; i < SPACELIMIT; ++i) 
                for (int j = 0; j < SPACELIMIT; ++j)
                    for (int k = 0; k < SPACELIMIT; ++k)
                        space2[a][i][j][k] = space[a][i][j][k];

        for (int a = 0; a < SPACELIMIT; ++a) {
            for (int i = 0; i < SPACELIMIT; ++i) {
                for (int j = 0; j < SPACELIMIT; ++j) {
                    for (int k = 0; k < SPACELIMIT; ++k) {
                        if (space[a][i][j][k] == '#' && !(count_around(space2, a, i, j, k) == 2 || count_around(space2, a, i, j, k) == 3))
                            space[a][i][j][k] = '.';
                        if (space[a][i][j][k] == '.' && count_around(space2, a, i, j, k) == 3)
                            space[a][i][j][k] = '#';
                    }
                }
            }
        }
    }

    int count = 0;
    for (int a = 0; a < SPACELIMIT; ++a)
        for (int i = 0; i < SPACELIMIT; ++i) 
            for (int j = 0; j < SPACELIMIT; ++j)
                for (int k = 0; k < SPACELIMIT; ++k)
                    if (space[a][i][j][k] == '#')
                        ++count;
    cout << count << endl;

    return 0;
}

int count_around (char space[SPACELIMIT][SPACELIMIT][SPACELIMIT][SPACELIMIT], int superslice, int slice, int row, int col) {

    // Outer edges of space (not indeded to be reached)
    if (superslice == 0 || superslice == SPACELIMIT - 1 || slice == 0 || slice == SPACELIMIT - 1 || row == 0 || row == SPACELIMIT - 1 || col == 0 || col == SPACELIMIT - 1)
        return 0;
    
    int count = 0;
    for (int a = superslice - 1; a < superslice + 2; ++a)
        for (int i = slice - 1; i < slice + 2; ++i)
            for (int j = row - 1; j < row + 2; ++j)
                for (int k = col - 1; k < col + 2; ++k)
                    if (space[a][i][j][k] == '#')
                        ++count;
                
    if (space[superslice][slice][row][col] == '#')
        --count;

    return count;
}